pkg/lease/lease_test.go

```diff
@@ -0,0 +1,305 @@
+	if srv.lastInterval != nil {
+		// The end timestamp must not go into the past.
+		if iv.EndTime.AsTime().Before(srv.lastInterval.EndTime.AsTime()) {
+			return nil, status.Error(codes.InvalidArgument, "end time before last end time")
+		}
+		// The frequency limit must at the very least prevent equal end timestamps.
+		if iv.EndTime.AsTime().Equal(srv.lastInterval.EndTime.AsTime()) {
+			return nil, status.Error(codes.InvalidArgument, "sample frequency too high")
+		}
+		if iv.EndTime.AsTime().Before(srv.lastInterval.EndTime.AsTime().Add(srv.frequencyLimit)) {
+			return nil, status.Error(codes.InvalidArgument, "sample frequency too high")
+		}
+		// Start time must not be before the last start time.
+		if iv.StartTime.AsTime().Before(srv.lastInterval.StartTime.AsTime()) {
+			return nil, status.Error(codes.InvalidArgument, "start time before last start time")
+		}
+	}
+	srv.lastInterval = iv
+
+	return &empty_pb.Empty{}, nil
+}
+
+type interval struct {
+	baseTime time.Time
+	replica  string
+
+	start, end time.Time
+}
+
+func (iv interval) String() string {
+	return fmt.Sprintf("%s[%s, %s]", iv.replica, iv.start.Sub(iv.baseTime), iv.end.Sub(iv.baseTime))
+}
+
+type replicaIntervals struct {
+	leaseOpts *Options
+	baseTime  time.Time
+
+	mtx sync.Mutex
+	m   map[string][]interval
+}
+
+func newReplicaIntervals(leaseOpts *Options, baseTime time.Time) *replicaIntervals {
+	return &replicaIntervals{
+		leaseOpts: leaseOpts,
+		baseTime:  baseTime,
+		m:         map[string][]interval{},
+	}
+}
+
+func (ri *replicaIntervals) add(replica string, start, end time.Time) {
+	ri.mtx.Lock()
+	defer ri.mtx.Unlock()
+
+	intervals := ri.m[replica]
+	// Do a replace if an interval is merely extended.
+	if last := len(intervals) - 1; last >= 0 && intervals[last].start.Equal(start) {
+		intervals = intervals[:last]
+	}
+	ri.m[replica] = append(intervals, interval{
+		replica:  replica,
+		baseTime: ri.baseTime,
+		start:    start,
+		end:      end,
+	})
+}
+
+func (ri replicaIntervals) intervals() []interval {
+	ri.mtx.Lock()
+	var intervals []interval
+	for _, iv := range ri.m {
+		intervals = append(intervals, iv...)
+	}
+	ri.mtx.Unlock()
+
+	if len(intervals) == 0 {
+		return nil
+	}
+	sort.Slice(intervals, func(i, j int) bool {
+		return intervals[i].start.Before(intervals[j].start)
+	})
+	return intervals
+}
+
+func (ri *replicaIntervals) validate() error {
+	// Intervals must never overlap but not have large gaps either.
+	// Start with a dummy interval at the front to simplify comparison.
+	iv1 := interval{
+		replica:  "dummy",
+		baseTime: ri.baseTime,
+		// A very quickly acquired lease may reach up to 1ms into the past with an
+		// empty initial state, so account for that.
+		start: ri.baseTime.Add(-time.Millisecond),
+		end:   ri.baseTime.Add(-time.Millisecond),
+	}
+
+	for _, iv2 := range ri.intervals() {
+		// Intervals are closed and thus must strictly not overlap.
+		if !iv1.end.Before(iv2.start) && !iv2.end.Before(iv1.start) {
```

Comment:
This check reads a bit awkward to me since `ri.intervals()` returns a sorted list. Is there any reason we can't check simply:
```go
if iv2.start.Before(iv1.end) || iv2.start.Equal(iv1.end) {
```